abulynka/shop-be
# Task 9 (Backend For Frontend)

1. Task: https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task9-bff-elastic-beanstalk/task.md
2. Deploy: http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com
Test examples:
 - get all products
   be: https://rawukoru6g.execute-api.eu-west-1.amazonaws.com/dev
   bff: http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
- add new product:
  be: https://rawukoru6g.execute-api.eu-west-1.amazonaws.com/dev/products
        with body: { "title": "1", "description": "2", "price": 1, "count": 2 }
  bff: http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
        with body: { "title": "1", "description": "2", "price": 1, "count": 2 }
- urls for carts:
  be: http://abulynka-cart-api-test.eu-west-1.elasticbeanstalk.com/api/profile/cart
  bff: http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/profile/cart

Content of .env file for localhost testing:
products=https://rawukoru6g.execute-api.eu-west-1.amazonaws.com/dev
profile-cart=http://abulynka-cart-api-test.eu-west-1.elasticbeanstalk.com/api
port=8081

How to test cache:
 Get products list:
      http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
 Create new product:
      http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
      with body:
      { "title": "1", "description": "2", "price": 1, "count": 2 }
 Get products list - result shouldn’t have new product:
      http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/products
 Wait more than 2 minutes
 Get products list - result should have new product:
      http://abulynka-bff-api-develop.eu-west-1.elasticbeanstalk.com/products

3. Done 04.11.2021 / deadline 06.11.2021
    Score: 7 / 7

3 - A working and correct express application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the task 9.1
5 - The bff-service should be deployed with Elastic Beanstalk. The bff-service call should be redirected to the appropriate service : product-service or CART. The response from the bff-service should be the same as if product-service or CART services were called directly.

Additional (optional) tasks
+1 - Add a cache at the bff-service level for a request to the getProductsList function of the product-service. The cache should expire in 2 minutes.
    How to test:
        Get products list
        Create new product
        Get products list - result shouldn’t have new product
        Wait more than 2 minutes
        Get products list - result should have new product
+1 - Use NestJS to create bff-service instead of express
